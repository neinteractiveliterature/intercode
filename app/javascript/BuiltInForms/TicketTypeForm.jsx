import React from 'react';
import PropTypes from 'prop-types';
import NumberInput from 'react-number-input';
import BooleanInput from '../BuiltInFormControls/BooleanInput';
import BootstrapFormInput from '../BuiltInFormControls/BootstrapFormInput';
import ResourceForm from './ResourceForm';
import ScheduledValueEditor from '../BuiltInFormControls/ScheduledValueEditor';
import { ScheduledValuePropType } from '../ScheduledValuePropTypes';

const buildScheduledMoneyValueInput = (value, onChange) => {
  const processNumberInputChangeEvent = (event, newValue) => {
    const moneyValue = {
      ...(value || {}),
      fractional: newValue * 100.0,
    };

    onChange(moneyValue);
  };

  let dollarValue = null;

  if (value && value.fractional !== null) {
    dollarValue = (value.fractional / 100.0).toFixed(2);
  }

  return (
    <div className="input-group">
      <span className="input-group-addon">$</span>
      <NumberInput
        className="form-control"
        value={dollarValue}
        onChange={processNumberInputChangeEvent}
        format="0,0.00"
      />
    </div>
  );
};

class TicketTypeForm extends React.Component {
  static propTypes = {
    initialTicketType: PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      pricing_schedule: ScheduledValuePropType.isRequired,
      publicly_available: PropTypes.bool.isRequired,
    }).isRequired,
    baseUrl: PropTypes.string.isRequired,
    timezone: PropTypes.string.isRequired,
  }

  constructor(props) {
    super(props);

    this.state = {
      ticketType: this.props.initialTicketType,
    };
  }

  getSubmitRequestBody = () => ({
    ticket_type: this.state.ticketType,
  })

  inputChanged = (event) => {
    this.setState({
      ticketType: {
        ...this.state.ticketType,
        [event.target.name]: event.target.value,
      },
    });
  }

  pricingScheduleChanged = (newPricingSchedule) => {
    this.setState({
      ticketType: {
        ...this.state.ticketType,
        pricing_schedule: newPricingSchedule,
      },
    });
  }

  publiclyAvailableChanged = (value) => {
    this.setState({
      ticketType: {
        ...this.state.ticketType,
        publicly_available: value,
      },
    });
  }

  render = () => {
    const disableSubmit = !ScheduledValueEditor.isValid(this.state.ticketType.pricing_schedule);

    return (
      <ResourceForm
        baseUrl={this.props.baseUrl}
        resourceId={this.state.ticketType.id}
        getSubmitRequestBody={this.getSubmitRequestBody}
        submitText="Save ticket type configuration"
        submitDisabled={disableSubmit}
      >
        <BootstrapFormInput
          label="Name (no spaces allowed &mdash; only letters, numbers, and underscores)"
          name="name"
          type="text"
          style={{ fontFamily: 'monospace' }}
          value={this.state.ticketType.name}
          onChange={this.inputChanged}
        />

        <BootstrapFormInput
          label="Description"
          name="description"
          type="text"
          value={this.state.ticketType.description}
          onChange={this.inputChanged}
        />

        <BooleanInput
          caption="Publicly available for purchase?"
          name="publicly_available"
          value={this.state.ticketType.publicly_available}
          onChange={this.publiclyAvailableChanged}
        />

        <fieldset>
          <p>Pricing schedule</p>

          <ScheduledValueEditor
            timezone={this.props.timezone}
            scheduledValue={this.state.ticketType.pricing_schedule}
            setScheduledValue={this.pricingScheduleChanged}
            buildValueInput={buildScheduledMoneyValueInput}
          />
        </fieldset>
      </ResourceForm>
    );
  }
}

export default TicketTypeForm;
