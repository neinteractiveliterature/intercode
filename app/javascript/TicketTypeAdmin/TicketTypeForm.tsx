import * as React from 'react';
import { pluralize } from 'inflected';
import {
  BooleanInput,
  BootstrapFormInput,
  parseIntOrNull,
  usePropertySetters,
} from '@neinteractiveliterature/litform';

import { TicketTypeAdmin_TicketTypeFieldsFragment } from './queries.generated';

export type EditingTicketType = Omit<
  TicketTypeAdmin_TicketTypeFieldsFragment,
  'maximum_event_provided_tickets'
> & {
  maximum_event_provided_tickets?: number | null;
};

export type TicketTypeFormProps = {
  ticketType: EditingTicketType;
  onChange: React.Dispatch<React.SetStateAction<EditingTicketType>>;
  ticketName: string;
};

function TicketTypeForm({ ticketType, ticketName, onChange }: TicketTypeFormProps) {
  const [
    setName,
    setDescription,
    setCountsTowardsConventionMaximum,
    setAllowsEventSignups,
    setMaximumEventProvidedTickets,
  ] = usePropertySetters(
    onChange,
    'name',
    'description',
    'counts_towards_convention_maximum',
    'allows_event_signups',
    'maximum_event_provided_tickets',
  );

  return (
    <div>
      <BootstrapFormInput
        label="Name (no spaces allowed &mdash; only letters, numbers, and underscores)"
        name="name"
        type="text"
        style={{ fontFamily: 'monospace' }}
        value={ticketType.name}
        onTextChange={setName}
      />

      <BootstrapFormInput
        label="Description"
        name="description"
        type="text"
        value={ticketType.description ?? ''}
        onTextChange={setDescription}
      />

      <BooleanInput
        caption="Allows event signups?"
        name="allows_event_signups"
        value={ticketType.allows_event_signups}
        onChange={setAllowsEventSignups}
      />

      <BooleanInput
        caption="Counts towards convention maximum?"
        name="counts_towards_convention_maximum"
        value={ticketType.counts_towards_convention_maximum}
        onChange={setCountsTowardsConventionMaximum}
      />

      <BootstrapFormInput
        label={`Number of event-provided ${pluralize(ticketName)} of this type (per event)`}
        name="maximum_event_provided_tickets"
        type="number"
        value={(ticketType.maximum_event_provided_tickets ?? '').toString()}
        onTextChange={(newValue) => setMaximumEventProvidedTickets(parseIntOrNull(newValue))}
      />
    </div>
  );
}

export default TicketTypeForm;
