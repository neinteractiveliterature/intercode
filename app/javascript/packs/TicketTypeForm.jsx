import React from 'react';
import PropTypes from 'prop-types';
import ScheduledValueEditor from './ScheduledValueEditor';
import { ScheduledValuePropType } from './ScheduledValuePropTypes';
import { fetchAndThrow, buildJSONFetchOptions } from './HTTPUtils';

class TicketTypeForm extends React.Component {
  static propTypes = {
    initialTicketType: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      pricing_schedule: ScheduledValuePropType.isRequired,
    }).isRequired,
    baseUrl: PropTypes.string.isRequired,
    timezone: PropTypes.string.isRequired,
  }

  constructor(props) {
    super(props);

    this.state = {
      ticketType: this.props.initialTicketType,
    };
  }

  setTicketTypeAttribute = (attribute, value) => {
    this.setState({ ticketType: Object.assign({}, this.state.ticketType, { [attribute]: value }) });
  }

  inputChanged = (event) => {
    const { name, value } = event.target;
    this.setTicketTypeAttribute(name, value);
  }

  pricingScheduleChanged = (newPricingSchedule) => {
    this.setTicketTypeAttribute('pricing_schedule', newPricingSchedule);
  }

  resourceUrl = () => {
    if (this.state.ticketType.id) {
      return `${this.props.baseUrl}/${this.state.ticketType.id}`;
    }

    return this.props.baseUrl;
  }

  isUpdate = () => this.state.ticketType.id

  submitClicked = (e) => {
    e.preventDefault();

    fetchAndThrow(this.resourceUrl(), buildJSONFetchOptions({
      method: (this.isUpdate() ? 'PATCH' : 'POST'),
      body: {
        ticket_type: this.state.ticketType,
      },
    })).then(() => {
      window.location.href = this.props.baseUrl;
    }).catch((error) => {
      alert(error.response.data);
    });
  }

  render = () => {
    const disableSubmit = !ScheduledValueEditor.isValid(this.state.ticketType.pricing_schedule);

    return (
      <form>
        <div className="form-group">
          <label htmlFor="name">
            Name (no spaces allowed &mdash; only letters, numbers, and underscores)
          </label>
          <input
            id="name"
            name="name"
            type="text"
            className="form-control"
            style={{ fontFamily: 'monospace' }}
            value={this.state.ticketType.name}
            onChange={this.inputChanged}
          />
        </div>

        <div className="form-group">
          <label htmlFor="description">Description</label>
          <input
            id="description"
            name="description"
            type="text"
            className="form-control"
            value={this.state.ticketType.description}
            onChange={this.inputChanged}
          />
        </div>

        <fieldset>
          <p>Pricing schedule</p>

          <ScheduledValueEditor
            timezone={this.props.timezone}
            scheduledValue={this.state.ticketType.pricing_schedule}
            setScheduledValue={this.pricingScheduleChanged}
          />
        </fieldset>

        <div className="form-group">
          <input
            type="submit"
            disabled={disableSubmit}
            onClick={this.submitClicked}
            className="btn btn-primary"
            value="Save ticket type configuration"
          />
        </div>
      </form>
    );
  }
}

export default TicketTypeForm;
