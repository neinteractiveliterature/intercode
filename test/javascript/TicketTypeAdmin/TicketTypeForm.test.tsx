import { render } from '../testUtils';
import TicketTypeForm from '../../../app/javascript/TicketTypeAdmin/TicketTypeForm';

test('TicketTypeForm matches snapshot', async () => {
  const { asFragment } = await render(
    <TicketTypeForm
      ticketType={{
        __typename: 'TicketType',
        allows_event_signups: true,
        id: '0',
        providing_products: [],
        name: '',
        description: '',
        maximum_event_provided_tickets: 0,
        counts_towards_convention_maximum: true,
      }}
      onChange={() => {}}
    />,
    {
      appRootContextValue: { ticketName: 'banana', ticketNamePlural: 'bananas' },
    },
  );

  expect(asFragment()).toMatchSnapshot();
});
