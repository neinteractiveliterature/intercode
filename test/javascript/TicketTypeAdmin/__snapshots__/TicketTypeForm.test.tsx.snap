// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketTypeForm matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="name-1"
      >
        Name (no spaces allowed â€” only letters, numbers, and underscores)
      </label>
      <input
        class="form-control"
        id="name-1"
        style="font-family: monospace;"
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="description-2"
      >
        Description
      </label>
      <input
        class="form-control"
        id="description-2"
        type="text"
        value=""
      />
    </div>
    <fieldset
      class="mb-3"
    >
      <legend
        class="col-form-label"
      >
        Allows event signups?
      </legend>
      <div>
        <div
          class="form-check form-check-inline"
        >
          <label
            class="form-check-label"
            for="allows_event_signups-3"
          >
            <input
              checked=""
              class="form-check-input"
              id="allows_event_signups-3"
              name="allows_event_signups"
              type="radio"
              value="true"
            />
             Yes
          </label>
        </div>
        <div
          class="form-check form-check-inline"
        >
          <label
            class="form-check-label"
            for="allows_event_signups-4"
          >
            <input
              class="form-check-input"
              id="allows_event_signups-4"
              name="allows_event_signups"
              type="radio"
              value="false"
            />
             No
          </label>
        </div>
      </div>
    </fieldset>
    <fieldset
      class="mb-3"
    >
      <legend
        class="col-form-label"
      >
        Counts towards convention maximum?
      </legend>
      <div>
        <div
          class="form-check form-check-inline"
        >
          <label
            class="form-check-label"
            for="counts_towards_convention_maximum-5"
          >
            <input
              checked=""
              class="form-check-input"
              id="counts_towards_convention_maximum-5"
              name="counts_towards_convention_maximum"
              type="radio"
              value="true"
            />
             Yes
          </label>
        </div>
        <div
          class="form-check form-check-inline"
        >
          <label
            class="form-check-label"
            for="counts_towards_convention_maximum-6"
          >
            <input
              class="form-check-input"
              id="counts_towards_convention_maximum-6"
              name="counts_towards_convention_maximum"
              type="radio"
              value="false"
            />
             No
          </label>
        </div>
      </div>
    </fieldset>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="maximum_event_provided_tickets-7"
      >
        Number of event-provided bananas of this type (per event)
      </label>
      <input
        class="form-control"
        id="maximum_event_provided_tickets-7"
        type="number"
        value="0"
      />
    </div>
  </div>
  <div>
    <div
      aria-hidden="true"
      class="modal fade"
      role="dialog"
      style="display: none; z-index: 1041;"
      tabindex="-1"
    >
      <div
        class="modal-dialog"
        role="document"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h5
              class="modal-title"
            >
              Confirmation
            </h5>
          </div>
          <div
            class="modal-body"
          >
            <div />
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-secondary"
              type="button"
            >
              Cancel
            </button>
            <button
              class="btn btn-primary"
              type="button"
            >
              OK
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
